### PR TITLE
Php templates

### PR DESCRIPTION
**Pull request by weaverryan**
This depends on #59. So sorry for the messy diff until that is merged:

tl;dr; This renders the templates as PHP, instead of simple `{{ foo }}` find and replace. We do not want to make the templates super dynamic and fancy, but we're already running into a few spots where we're hitting this limitation (case: `make:controller` which has *two* possible templates, because the current replace is too limiting).

The only ugly part is that the `<?php` in the templates needs to be written as `//PHP_OPEN` to avoid syntax errors (we then replace this with `<?php` after including). Here is an example: https://github.com/symfony/maker-bundle/compare/master...weaverryan:php-templates?expand=1#diff-3ecc33834a0f1ec4f8e5516d71e4ad35

Btw, if you're wondering "Why not use Twig!?". We don't want to create a dependency on Twig to use the maker :).

Cheers!

### CODE REVIEW

File: src/Resources/skeleton/authenticator/Empty.tpl.php [hunk 1]
```diff
@@ -9,7 +9,7 @@
 use Symfony\Component\Security\Core\User\UserProviderInterface;
 use Symfony\Component\Security\Guard\GuardAuthenticator;
 
-class {{ class_name }} extends GuardAuthenticator
+class <?php echo $class_name; ?> extends GuardAuthenticator
```
Comment (javiereguiluz) [inline src/Resources/skeleton/authenticator/Empty.tpl.php]: Just asking: would it be acceptable to replace this:

```php
<?php echo $var; ?>
```

by this?

```php
<?= $var ?>
```

The `<?=` syntax is always available since PHP 5.4 and it cannot be disabled in any way.

Comment (weaverryan) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: Hmm, I didn't know that. I thought that syntax was somehow deemed "horrible" by someone... but I can't remember. I'm fine with it - we'll see if anyone has a good reason against it.

Comment (javiereguiluz) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: Maybe it's horrible. Not sure. The `<?php echo ...` looked too long to me ... but maybe it's better to use it because it's the normal syntax everybody know and likes.

Comment (nicolas-grekas) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: I like the short syntax :) This is PHP, let's use the language's features ;)

File: src/DependencyInjection/CompilerPass/MakeCommandRegistrationPass.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+use Symfony\Bundle\MakerBundle\MakerInterface;
+use Symfony\Bundle\MakerBundle\Str;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Reference;
+
+class MakeCommandRegistrationPass implements CompilerPassInterface
+{
+    const MAKER_TAG = 'maker.command';
```
Comment (piotrgradzinski) [inline src/DependencyInjection/CompilerPass/MakeCommandRegistrationPass.php]: Shouldn't we use explicit `public const`?

Comment (weaverryan) [reply src/DependencyInjection/CompilerPass/MakeCommandRegistrationPass.php]: Why not? :) I've done it over on #59

**Comment by weaverryan** [comment]
This is rebased and ready. I'm still getting false fabbot failures on the `[]` array syntax...

File: src/FileManager.php [hunk 3]
```diff
@@ -40,13 +40,15 @@ public function setIO(SymfonyStyle $io): void
...
     {
-        $keys = array_keys($parameters);
-        $values = array_values($parameters);
-        $placeholders = array_map(function ($name) {
-            return "{{ $name }}";
-        }, $keys);
+        ob_start();
+        extract($parameters);
+        include($templatePath);
+        $contents = ob_get_contents();
```
Comment (nicolas-grekas) [inline src/FileManager.php]: ob_get_clean()

File: src/FileManager.php [hunk 4]
```diff
@@ -40,13 +40,15 @@ public function setIO(SymfonyStyle $io): void
...
+        ob_end_clean();
 
-        return str_replace($placeholders, $values, file_get_contents($templatePath));
+        $contents = str_replace('//PHP_OPEN', '<?php', $contents);
```
Comment (nicolas-grekas) [inline src/FileManager.php]: return str_replace...?

File: src/Resources/skeleton/authenticator/Empty.tpl.php [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<?php
+//PHP_OPEN
```
Comment (nicolas-grekas) [inline src/Resources/skeleton/authenticator/Empty.tpl.php]: `<?php ?>` instead?

Comment (weaverryan) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: But then it would be evaluated by the PHP processor, and would disappear. I guess we could *always* assume we need to prepend the entire file with `<?php`. Is it worth it? They're bother weird honestly...

**Comment by weaverryan** [comment]
Changes made... except for the one change...

File: src/Resources/skeleton/authenticator/Empty.tpl.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: > we could always assume we need to prepend the entire file

looks like a good idea to me

File: src/Resources/skeleton/authenticator/Empty.tpl.php (hunk 1 above)
Comment (fabpot) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: I like the short syntax as well.

File: src/Resources/skeleton/authenticator/Empty.tpl.php (hunk 5 above)
Comment (stof) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: always prepending is bad IMO: a generator may have a case for generating another type of file.

another option might be this (if it works):

```php
<?= '<?php' ?>
```

File: src/FileManager.php [hunk 6]
```diff
@@ -40,13 +40,12 @@ public function setIO(SymfonyStyle $io): void
 
     public function parseTemplate(string $templatePath, array $parameters): string
     {
-        $keys = array_keys($parameters);
-        $values = array_values($parameters);
-        $placeholders = array_map(function ($name) {
-            return "{{ $name }}";
-        }, $keys);
+        ob_start();
+        extract($parameters);
```
Comment (stof) [inline src/FileManager.php]: should use `EXTR_SKIP` instead of the default `EXTR_OVERWRITE`

File: src/Resources/skeleton/event/Subscriber.tpl.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
+//PHP_OPEN
+
+namespace App\EventSubscriber;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+<?php echo $eventUseStatement; ?>
+class <?php echo $subscriber_class_name; ?> implements EventSubscriberInterface
```
Comment (stof) [inline src/Resources/skeleton/event/Subscriber.tpl.php]: should have an empty line between the use statement and the class

File: src/Resources/skeleton/authenticator/Empty.tpl.php (hunk 1 above)
Comment (stof) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: I agree about using the short echo syntax

File: src/Resources/skeleton/authenticator/Empty.tpl.php (hunk 5 above)
Comment (weaverryan) [reply src/Resources/skeleton/authenticator/Empty.tpl.php]: What about `//<?php`?

**Comment by weaverryan** [comment]
Changes made! fabbot still dislikes me, but it's unrelated :)

**Comment by stof** [comment]
@fabpot any idea why fabbot would not use the config file in the repo ?

File: src/MakerInterface.php [hunk 8]
```diff
@@ -71,9 +71,13 @@ public function getParameters(InputInterface $input): array;
      * For example:
      *
      *    return array(
-     *        __DIR__.'/../Resources/skeleton/command/Command.php.txt' => 'src/Command/'.$params['command_class_name'].'.php',
+     *        __DIR__.'/../Resources/skeleton/command/Command.tpl.php' => 'src/Command/'.$params['command_class_name'].'.php',
      *    );
      *
+     * These files are parsed as PHP. To avoid syntax errors, replace the
+     * <?php tag in your templates with `//PHP_OPEN`. After evaluating the
```
Comment (nicolas-grekas) [inline src/MakerInterface.php]: leftover

File: src/Resources/skeleton/authenticator/Empty.tpl.php [hunk 9]
```diff
@@ -1,4 +1,5 @@
-<?php
+<?= '<?php' ?>
```
Comment (nicolas-grekas) [inline src/Resources/skeleton/authenticator/Empty.tpl.php]: `<?= "<?php\n" ?>`?

File: src/Resources/skeleton/event/Subscriber.tpl.php [hunk 10]
```diff
@@ -0,0 +1,22 @@
+<?= '<?php' ?>
+
+
+namespace App\EventSubscriber;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+<?= $eventUseStatement ?>
+
```
Comment (stof) [inline src/Resources/skeleton/event/Subscriber.tpl.php]: I think you need a second empty line here: The first EOL char will be swallowed by `?>`, so this would not create an empty line between the use statement and the class at the end.

**Comment by weaverryan** [comment]
@stof nice catch! I've just added php-cs-fixer to the tests, so that we can verify the generated files follow coding standards (especially important because that empty line stuff is very easy to mess up).

Thanks!
